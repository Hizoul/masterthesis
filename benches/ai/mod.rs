pub mod minmax;
pub mod mcts;
pub mod unsafe_mcts;