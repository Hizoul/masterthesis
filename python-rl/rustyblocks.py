#!/usr/bin/env python3
import math
import sys, ctypes
from os.path import abspath, dirname, realpath
from ctypes import c_char_p, c_uint32, c_uint8, c_int8, Structure, POINTER,c_void_p,c_int64,c_int16,c_double,c_uint64,c_char_p
import numpy as np
from numpy.ctypeslib import ndpointer
from PIL import Image


parent_folder = dirname(realpath(__file__)) + "/"
class Tuple(Structure):
    _fields_ = [("placed", c_uint8),("reward", c_double),("done", c_uint8),("winner", c_uint8)]
    def __str__(self):
        return "({},{},{},{})".format(self.placed, self.reward,self.done,self.winner)

class GameFieldS(Structure):
    pass

# libpath = "../target/debug/"
libpath = "../target/release/"
libname = "rustyblocks"
prefix = {'win32': ''}.get(sys.platform, 'lib')
extension = {'darwin': '.dylib', 'win32': '.dll'}.get(sys.platform, '.so')
absolute_lib_path = abspath(parent_folder + libpath + prefix + libname + extension)
print("Loading rustyblocks from ", absolute_lib_path)
rustLib = ctypes.cdll.LoadLibrary(absolute_lib_path)

rustLib.field_new.restype = POINTER(GameFieldS)
rustLib.field_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_new.restype = POINTER(GameFieldS)
rustLib.field_clone.argtypes = (POINTER(GameFieldS), )
rustLib.field_clone.restype = POINTER(GameFieldS)
rustLib.hex_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.hex_do_action_with_answer.restype = Tuple
rustLib.hex_to_array.restype = ndpointer(dtype=ctypes.c_uint8, shape=(121,))
rustLib.free_hex_array.argtypes = (ndpointer(dtype=ctypes.c_uint8, shape=(121,)), )
rustLib.hex_reset.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.restype = c_void_p
rustLib.field_to_json_log.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_json_log.restype = c_void_p
rustLib.field_restore_log.argtypes = (POINTER(GameFieldS), c_char_p)
rustLib.field_to_array.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_to_array.restype = ndpointer(dtype=c_int8, shape=(420,))
rustLib.free_field_array.argtypes = (ndpointer(dtype=c_int8, shape=(420,)), )
rustLib.field_get_score.argtypes = (POINTER(GameFieldS),)
rustLib.field_get_score.restype = ndpointer(dtype=c_int16, shape=(2,))
rustLib.free_score_array.argtypes = (ndpointer(dtype=c_int16, shape=(2,)), )
rustLib.field_do_action.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_do_action.restype = c_uint8
rustLib.field_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.field_do_action_with_answer.restype = Tuple
rustLib.field_do_action_with_answer_heu.argtypes = (POINTER(GameFieldS), c_double,c_double,c_double,c_double, c_uint8)
rustLib.field_do_action_with_answer_heu.restype = Tuple
rustLib.field_do_action_self_play.argtypes = (POINTER(GameFieldS), c_uint8, c_int8)
rustLib.field_do_action_self_play.restype = Tuple
rustLib.field_is_game_over.argtypes = (POINTER(GameFieldS), )
rustLib.field_is_game_over.restype = c_uint8
rustLib.field_get_reward.argtypes = (POINTER(GameFieldS), )
rustLib.field_get_reward.restype = c_double
rustLib.field_counter_action.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action.restype = c_int64
rustLib.field_counter_action_index.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action_index.restype = c_uint8
rustLib.field_reset.argtypes = (POINTER(GameFieldS), )
rustLib.free_string.argtypes = (c_void_p, )
rustLib.eval_heuristic_weights.argtypes = (c_double,c_double,c_double,c_double)
rustLib.eval_heuristic_weights.restype = c_uint64

def field_to_string(field):
  ptr = rustLib.field_to_string(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)
def field_to_log(field):
  ptr = rustLib.field_to_json_log(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)

def field_to_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy().reshape((42,10))
  rustLib.free_field_array(original_array)
  return result

def field_score(field):
  original_array = rustLib.field_get_score(field)
  result = original_array.copy()
  rustLib.free_score_array(original_array)
  return result

def field_to_cnn_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy()
  newres = np.zeros(shape=(32,32,3), dtype=np.uint8)
  for i in range(0, 200):
    y = math.floor(i / 10)
    x = i % 10
    current = 0
    val = 0
    if result[i] == 1 or result[i] == 2:
      current = 0
      if result[i] == 1:
        val = 128
      else:
        val = 255
    elif result[i] == 3 or result[i] == 4:
      current = 2
      if result[i] == 3:
        val = 128
      else:
        val = 255
    newres[y][x][current] = val
  for i in range(200, 210):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i] + 124
    newres[y][x][1] = val
  for i in range(210, 230):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    elif val > 0:
      val = val * 50
    newres[y][x][1] = val
  for i in range(230, 420):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    else:
      val = 255
    newres[y][x][1] = val
  rustLib.free_field_array(original_array)
  return newres

def hex_to_array(field):
  original_array = rustLib.hex_to_array(field)
  result = original_array.copy().reshape((11,11))
  rustLib.free_hex_array(original_array)
  return result

def field_string_to_array(field_string):
  field_array = np.zeros((20,10), dtype=np.uint8)
  row = 0
  col = 0
  for i in range(len(field_string)):
    entry = field_string[i]
    if entry == '\n':
      col = 0
      row += 1
    elif row < 20 and col < 10:
      field_array[row][col] = ord(entry) - 48
      col += 1
  return field_array

def save_img(data, name):
  Image.fromarray(data, 'RGB').save(name)