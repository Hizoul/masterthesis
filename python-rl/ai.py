import sys
import gym
from stable_baselines.common.policies import MlpPolicy
from stable_baselines import PPO2
from stable_baselines.common.vec_env import DummyVecEnv,VecCheckNan
import numpy as np
from os import listdir
from os.path import isfile, join
from stable_baselines.gail import ExpertDataset
from generate_pretraindata import generate_pretraindata_heuristics
from rustyblocks import rustLib, field_to_array, field_to_log

MAX_FOR_BLOCKS = 230
botToUse = sys.argv[1]
# oldLog = sys.argv[2]
oldLog = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":3,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}"

is_box_space = True
dirname = "D:\\4-System\\rusty\\"
filename="50000_heutistic_pretrain_"
filename += "box" if is_box_space else "discrete"
np.seterr(all='raise')

origEnv = gym.make("rustybox-v0" if is_box_space else "rustydiscrete-v0")

origEnv.max_invalid_tries = 7
env = VecCheckNan(DummyVecEnv([lambda: origEnv]))

# Instantiate the agent
model = PPO2.load("models/ppo2boxbestparam/2e4-30.pkl", env=env)
# model.load("models/pretrain/"+filename)

rustLib.field_restore_log(origEnv.field, oldLog.encode('utf-8'))
obs = field_to_array(origEnv.field)
actions, _states = model.predict(obs)
if is_box_space:
  action = 0
  current_max = -1
  for i in range(0, len(actions)):
    action_probability = actions[i]
    newIndex = MAX_FOR_BLOCKS + i
    action_possible = obs[int(newIndex / 10)][newIndex % 10] == 1
    if action_probability > current_max and action_possible:
      current_max = action_probability
      action = i
    rustLib.field_do_action(origEnv.field, action)
else:
  rustLib.field_do_action(origEnv.field, actions)
print("1;"+field_to_log(origEnv.field))