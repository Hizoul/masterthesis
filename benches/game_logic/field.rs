extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::game_logic::cache::{ShapeCache};
use std::sync::{Arc};
use rand::prelude::*;

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


struct Foo {
  counter: usize
}

impl Foo {
  pub fn add(&mut self) {self.counter+=1;}
  pub fn get(&self) -> usize {self.counter}
}

trait FooTrait {
  fn t_add(&mut self);
  fn t_get(&self) -> usize;
}

impl FooTrait for Foo {
  fn t_add(&mut self) {self.add();}
  fn t_get(&self) -> usize {self.get()}
}

pub fn bench_field(c: &mut Criterion) {
  let cache = ShapeCache::new();
  let rules = GameRules::default();
  let p = Arc::new(cache);
  let p2 = p.clone();
  let p3 = p.clone();
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  c.bench_function("deterministic_play_until_over", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p.clone(), rules);
    while !f.game_over {
      f.place_block_using_play(0);
    }
  }));
  c.bench_function("random_play_until_over", move |b| b.iter(|| {
    let mut  f = GameField::new_with_cache(2, p2.clone(), rules);
    while !f.game_over {
      let p = f.possible_plays.len() as u8;
      f.place_block_using_play(rand::thread_rng().gen_range(0,p));
    }
  }));
  c.bench_function("replay_log", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p3.clone(), rules);
    f.restore_from_log(&log, false);
  }));
  c.bench_function("traitcost_direct", move |b| b.iter(|| {
    let mut f = Foo {counter:0};
    for _ in 0..99999999 {
      while f.get() < 9999999999 {
        f.add();
      }
    }
  }));
  c.bench_function("traitcost_struct", move |b| b.iter(|| {
    let mut f: Box<dyn FooTrait> = Box::new(Foo {counter:0});
    for _ in 0..99999999 {
      while f.t_get() < 9999999999 {
        f.t_add();
      }
    }
  }));
}