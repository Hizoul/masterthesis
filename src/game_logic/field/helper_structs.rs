use serde::{Deserialize, Serialize};
use crate::game_logic::tetromino::{Tetromino};
use std::sync::{Arc};

#[derive(Debug,Clone)]
pub struct Touching {
  pub block: Arc<Tetromino>,
  pub touching: Vec<Arc<Tetromino>>,
  pub connected: Vec<Arc<Tetromino>>
}

impl PartialEq for Touching {
  fn eq(&self, other: &Touching) -> bool {
    self.block == other.block
  }
}

impl Touching {
  pub fn new(block: Arc<Tetromino>) -> Touching {
    return Touching {
      block, touching: Vec::with_capacity(4), connected: Vec::with_capacity(4)
    }
  }
}

#[derive(Debug,Serialize,Deserialize,Clone)]
pub struct OwnedPosition {
    pub x: i8,
    pub y: i8,
    pub from: i8
}

#[derive(Debug,Serialize,Deserialize,Copy,Clone)]
pub struct GameRules {
  pub roll_dice: bool,
  pub limit_minus: bool
}

impl GameRules {
  pub fn original() -> GameRules {
    GameRules {roll_dice:true,limit_minus:true}
  }
  pub fn original_without_limit() -> GameRules {
    GameRules {roll_dice:true,limit_minus:false}
  }
  pub fn deterministic() -> GameRules {
    GameRules {roll_dice:false,limit_minus:true}
  }
  pub fn deterministic_without_limit() -> GameRules {
    GameRules {roll_dice:false,limit_minus:false}
  }
}

impl Default for GameRules {
  fn default() -> GameRules {
    GameRules::original_without_limit()
  }
}