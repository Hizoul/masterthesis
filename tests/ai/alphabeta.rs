use rustyblocks::ai::alphabeta::{AlphaBetaBot};
use rustyblocks::game_player::{PlaySupervisor};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::field::render::Renderable;
use insta::{assert_snapshot};

#[test]
fn ai_alphabeta() {
  let b1 = AlphaBetaBot::new(0, 4);
  let b2 = AlphaBetaBot::new(1, 1);
  let mut ps = PlaySupervisor::new_with_rules(vec!(Box::new(b1), Box::new(b2)), GameRules::deterministic());
  ps.play();
  assert_snapshot!("field_after_game", ps.field.to_field_string());
  assert_snapshot!("log_after_game", serde_json::to_string(&ps.field.log).unwrap());
}