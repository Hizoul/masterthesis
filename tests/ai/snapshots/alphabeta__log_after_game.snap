---
created: "2019-09-18T11:15:52.189989752Z"
creator: insta@0.8.1
source: tests/ai/alphabeta.rs
expression: "serde_json::to_string(&ps.field.log).unwrap()"
---
{"log":[{"PayloadPlaced":{"x":0,"y":0,"block":0,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":1,"block":0,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":2,"block":0,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":3,"block":0,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":4,"y":0,"block":0,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":4,"block":0,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":4,"y":4,"block":0,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":5,"block":0,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":6,"block":0,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":7,"block":0,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":3,"y":8,"block":1,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":8,"block":1,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":3,"y":10,"block":1,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":10,"block":1,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":3,"y":12,"block":1,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":12,"block":1,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":3,"y":14,"block":1,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":14,"block":1,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":3,"y":16,"block":1,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":16,"block":1,"orientation":0,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":6,"y":0,"block":2,"orientation":0,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":5,"y":2,"block":2,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":6,"y":4,"block":2,"orientation":3,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":5,"y":6,"block":2,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":6,"y":8,"block":2,"orientation":3,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":5,"y":10,"block":2,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":6,"y":12,"block":2,"orientation":3,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":5,"y":14,"block":2,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":6,"y":16,"block":2,"orientation":3,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":5,"y":18,"block":2,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":2,"block":3,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":4,"block":3,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":6,"block":3,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":8,"block":3,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":10,"block":3,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":12,"block":3,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":14,"block":3,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":16,"block":3,"orientation":1,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":7,"y":18,"block":3,"orientation":1,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":2,"y":17,"block":4,"orientation":2,"from":1}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":6,"y":19,"block":4,"orientation":6,"from":0}},{"PayloadConsidering":{"play_index":0}},{"PayloadPlaced":{"x":0,"y":18,"block":3,"orientation":0,"from":1}},{"PayloadSkipped":{"from":0,"block":3,"reason":1}},{"PayloadSkipped":{"from":1,"block":3,"reason":1}}]}
